File: Top_Interview_Questions/17_LetterCombinationsOfAPhoneNumber.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};
        vector<string> letters = {"",    "",    "abc",  "def", "ghi",
                                  "jkl", "mno", "pqrs", "tuv", "wxyz"};
        const int n = digits.size();
        vector<string> ans{""};
        for (int i = 0; i < n; ++i) {
            vector<string> t;
            string str = letters[digits[i] - '0'];
            for (const auto& j : str) {
                for (const auto& k : ans) t.push_back(k + j);
            }
            ans = t;
        }
        return ans;
    }
};

// dfs
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.empty()) return {};

        vector<string> ans;
        vector<string> letters = {"",    "",    "abc",  "def", "ghi",
                                  "jkl", "mno", "pqrs", "tuv", "wxyz"};
        string cur;
        dfs(digits, letters, 0, cur, ans);
        return ans;
    }
    void dfs(string& digits, vector<string>& letters, int loc, string& cur,
             vector<string>& ans) {
        if (loc == digits.size()) {
            ans.push_back(cur);
            return;
        }

        for (int i = 0; i < letters[digits[loc] - '0'].size(); ++i) {
            cur.push_back(letters[digits[loc] - '0'][i]);
            dfs(digits, letters, loc + 1, cur, ans);
            cur.pop_back();
        }
    }
};

File: Question_Go/17_Letter_Combinations_of_a_Phone_Number.go
func letterCombinations(digits string) []string {
    if digits == "" {
		return nil
	}
    letters := []string{"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs","tuv", "wxyz"}

    ans := make([]string, 0)
    dfs(digits, 0, letters, "", &ans)
    return ans
}

func dfs(digits string, loc int, letters []string, cur string, ans *[]string) {
    if loc == len(digits) {
        *ans = append(*ans, cur)
        return
    }

    for _, ch := range letters[int(digits[loc] - '0')] {
        dfs(digits, loc + 1, letters, cur + string(ch), ans)
    }
}